### ISSUE TITLE
Remove need for startup application to fix health_mode persistence

### ISSUE THREAD

**Issue by brenobaptista** [frederik-h/acer-wmi-battery#47]
I've described what I found out on the PR for DKMS support. But basically another Acer Nitro 5 owner and I found out that by creating a systemd service to force health_mode to be 1 on startup fixed the problem of it going back to 0 after a few days for no apparent reason.

I have no experience with building drivers but maybe to fix this it would be needed to persist this value in a file and get that file on reboot/kernel updates.

I made a quick guide on my personal blog explaining how to use this driver and it would be nice to remove the extra step, it's a big chunk of the whole guide: https://www.brenobaptista.com/posts/battery-charging-threshold-acer-linux

**Comment by frederik-h** [comment]
I am not aware of a a way to make the health mode setting persistent across reboots. I would also guess there is none. You could pass the  `enable_health_mode=1` argument to insmod when you load the module, e.g. from `/etc/modules.` This way you won't need an additional script for writing to `/sys/bus/wmi/drivers/acer-wmi-battery/health_mode`. I suppose you could also add the argument to [this file](https://github.com/frederik-h/acer-wmi-battery/pull/31/files#diff-a6b825bb9f177472e6d958e12d322174572a4e47e79d6a50339cd85098a80cc7) if you use the dkms setup, but I have not tried it.

**Comment by ryanbarillos** [comment]
I may chime in just a little bit, based on my experience making the installation package for this driver on Arch Linux
- As soon as I reboot (after installing either the [dkms](https://aur.archlinux.org/packages/acer-wmi-battery-dkms) or [dkms-git](https://aur.archlinux.org/packages/acer-wmi-battery-dkms-git) package), _health_mode is automatically set to 1_

Which is surprising to me—and it doesn't require extra scripts to force that 1 but a simple reboot. (You may git clone these packages & inspect the code, but it's mostly dry of any magic.)

Only did different was make an acer-wmi-battery.conf to `/usr/lib/modules-load.d/` as advised to me from the [Arch Linux forums](https://bbs.archlinux.org/viewtopic.php?pid=2168092#p2168092), but nothing else.

**Comment by frederik-h** [comment]
> I may chime in just a little bit, based on my experience making the installation package for this driver on Arch Linux
> 
> * As soon as I reboot (after installing either the [dkms](https://aur.archlinux.org/packages/acer-wmi-battery-dkms) or [dkms-git](https://aur.archlinux.org/packages/acer-wmi-battery-dkms-git) package of this driver), _health_mode is automatically set to 1_
> 
> Which is surprising to me—and it doesn't require extra scripts to force that 1 but a simple reboot. (You may git clone these packages & inspect the code, but it's mostly dry of any magic.)

Perhaps some Acer laptops persist the value and some do not. That seems totally possible to me. But there seems to be no way to control this in the WMI interface. On my laptop, the setting is gone after a cold reboot. Which model do you have?

**Comment by ryanbarillos** [comment]
> Perhaps some Acer laptops persist the value and some do not. That seems totally possible to me. But there seems to be no way to control this in the WMI interface. On my laptop, the setting is gone after a cold reboot. Which model do you have?

[Acer Spin 3 (SP314-21N)](https://github.com/frederik-h/acer-wmi-battery/issues/51), which was released on 2021.

And I did make sure to turn on Battery Charge Limit from Acer Control Center prior to migrating to Linux.

**Comment by ryanbarillos** [comment]
> But there seems to be no way to control this in the WMI interface.

From all I know on my specific model, such control only existed at the Acer Control Center on Windows 10/11 (the toggle was there briefly until it vanished from a reinstall of Windows few years ago).

Unless that program was somehow reverse-engineered to work on Linux, or implement an open-source framework space to control Acer firmware/BIOS values (such as for [Asus](https://asus-linux.org/) with [asusctl](https://wiki.archlinux.org/title/Asusctl)), I have doubts on the permanence on keeping Health Mode on with just a _sudo echo 1_ without it being overridden afterward.

**Comment by frederik-h** [comment]
> > But there seems to be no way to control this in the WMI interface.
> [...] 
> EDIT: Found this [GitHub issue from TLP](https://github.com/linrunner/TLP/issues/596) while browsing through @frederik-h , and it seems that you may be aware of this already.

This driver is based on reverse-engineering the Windows tool and uses the WMI interface. But I did not find anything that would let you persist the state of the "health mode" between cold reboots. It seems to survive soft reboots and waking up from hibernation on my laptop. If you want it to be enabled by default, you need some init script that enables the health mode after each boot or you can enable it by passing a flag to the module when you load it. The latter option seems a bit easier to me, by the way.

**Comment by ryanbarillos** [comment]
> If you want it to be enabled by default, you need some init script that enables the health mode after each boot or you can enable it by passing a flag to the module when you load it. The latter option seems a bit easier to me, by the way.

@frederik-h A SystemD service should work. Besides the one proposed by @brenobaptista, one idea is to use a loop to check & change the health mode value (or a system trigger, which I'm not aware of where to find).

**Comment by frederik-h** [comment]
> > If you want it to be enabled by default, you need some init script that enables the health mode after each boot or you can enable it by passing a flag to the module when you load it. The latter option seems a bit easier to me, by the way.
> 
> @frederik-h A SystemD service should work. Besides the one proposed by @brenobaptista, one idea is to use a loop (separated by an interval of few seconds) to check & change the health mode value, or a system trigger, (which I'm not aware of where to find).

There are lots of possibilities to enable the health mode on boot. You can use an init script or even a systemd service to write to `/sys/bus/wmi/drivers/acer-wmi-battery/health_mode`, but I would recommend to load the module, for instance, in `/etc/modules` and pass the `enable_health_mode=1` argument. There is no need to loop over anything. As I said, the value should never get reset. There was however a report about some problems caused by a fan control driver which would reset the health mode setting, cf. #40. Hence, if the health mode settings gets reset on your system, this is probably caused by a bug in another driver.

**Comment by ryanbarillos** [comment]
> Perhaps some Acer laptops persist the value and some do not. That seems totally possible to me. But there seems to be no way to control this in the WMI interface. On my laptop, the setting is gone after a cold reboot. Which model do you have?

This week, unfortunately, my health_mode no longer persists. Now I also have to set it to 1 manually every time I boot my laptop.

**Comment by psiwray** [comment]
Sorry for reviving an old thread, @ryanbarillosofficial. I just configured my health mode to be enabled on boot another way, it might be useful to you too (I'm on Ubuntu 24.10):

+ I have my kernel module copied into `/lib/modules/$(uname -r)/kernel/acer/acer-wmi-battery.ko`;
+ I created a file `/etc/modules-load.d/acer-wmi-battery.conf` with a single line: `acer-wmi-battery` to instruct the kernel to load this module on boot;
+ I created another file `/etc/modprobe.d/acer-wmi-battery.conf` with a single line: `options acer-wmi-battery enable_health_mode=1` to use that option when loading the kernel module;
+ I ran `sudo depmod -a` to update the kernel modules dependencies information.

After a reboot, doing `sudo cat /sys/bus/wmi/drivers/acer-wmi-battery/health_mode` outputs `1`. Isn't this your desired behavior? This persists across reboots.

**Comment by brenobaptista** [comment]
Thanks @psiwray!
From what I understand, on these instructions you're installing a manual kernel module (check https://github.com/Diman119/acer-wmi-battery for dkms) + loading the module with enable_health_mode.

I just needed the second part, as I'm using a repository that sets up akmods for me (using Fedora now): https://copr.fedorainfracloud.org/coprs/asan/acer-modules

So I removed my startup application and instead will try this method.

Basically, to fix this I just had to run:
```
echo "options acer_wmi_battery enable_health_mode=1" | sudo tee /etc/modprobe.d/acer-wmi-battery.conf
```
